storage/src/robocorp/storage/__init__.py

```diff
@@ -0,0 +1,173 @@
+import logging
+import random
+import time
+from typing import Dict, List, Optional
+
+from ._requests import Requests, RequestsHTTPError
+from ._storage import AssetNotFound
+from ._storage import get_assets_client as _get_assets_client
+from ._utils import with_lazy_objects as _with_lazy_objects
+
+__version__ = "0.1.0"
+version_info = [int(x) for x in __version__.split(".")]
+
+LOGGER = logging.getLogger(__name__)
+_with_lazy_assets = _with_lazy_objects(_assets_client=_get_assets_client)
+
+
+@_with_lazy_assets
+def list_assets(*, _assets_client: Requests) -> List[Dict]:
+    """List all the existing assets.
+
+    Returns:
+        A list of assets where each asset is a dictionary with fields like 'id' and
+        'name'.
+    """
+    assets = _assets_client.get("").json()
+    LOGGER.info("Found %d assets.", len(assets))
+    return assets
+
+
+def _retrieve_asset_id(name: str) -> str:
+    for asset in list_assets():
+        if asset["name"] == name:
+            asset_id = asset["id"]
+            LOGGER.debug("Found existing asset %r with ID: %s", name, asset_id)
+            return asset_id
+
+    LOGGER.warning("No asset with name %r found, assuming ID.", name)
+    return name
+
+
+@_with_lazy_assets
+def _get_asset(
+    name: str, *, _assets_client: Requests, raise_if_missing: bool = True
+) -> Optional[Dict]:
+    asset_id = _retrieve_asset_id(name)
+    exception = None
+
+    def _handle_error(resp):
+        # NOTE(cmin764): The API will return 404 if the asset can't be found on the
+        #  server, therefore don't let this raise, nor retry given this custom handler.
+        nonlocal exception
+        try:
+            _assets_client.handle_error(resp)
+        except RequestsHTTPError as exc:
+            if exc.status_code == 404:
+                exception = exc
+            else:
+                raise
+
+    LOGGER.debug("Retrieving asset %r with resulted ID %r.", name, asset_id)
+    response = _assets_client.get(asset_id, _handle_error=_handle_error)
+    if response.ok:
+        return response.json()
+
+    message = f"asset with name {name!r} and resulted ID {asset_id!r} couldn't be found"
+    if raise_if_missing:
+        raise AssetNotFound(message) from exception
+
+    LOGGER.warning("%s%s!", message[0].upper(), message[1:])
+    return None
+
+
+@_with_lazy_assets
+def get_asset(name: str, *, _assets_client: Requests) -> str:
+    """Get the asset's value by providing its `name`.
+
+    Args:
+        name: Name of the asset to fetch.
+
+    Returns:
+        The previously set value of this asset. (empty if it was never set)
+
+    Raises:
+        AssetNotFound: When the queried asset doesn't exist.
+    """
+    LOGGER.info("Retrieving asset %r.", name)
+    payload = _get_asset(name)["payload"]
+    if payload["type"] == "empty":
```

Comment:
Related to previous comment about types, these would also be easier to understand as a new developer coming in, if the API return types had some shape to them. But still, can be a later improvement.

Reply:
Yes, totally makes sense here what to expect when you have: `Dict[str, Union[str, Dict[str, str]]]`.

Reply:
You can also define more strict types for containers with a guaranteed structure. I think it would be something along these lines:
```python
from typing import TypedDict, Optional, cast

Payload = TypedDict("Payload", {
    "type": str,
    "content_type": Optional[str],
    "url": Optional[str]
})

Asset = TypedDict("Asset", {
    "id": str,
    "name": str,
    "payload": Payload,
})

def get_asset(name) -> Asset:
    return cast(Asset, _client.get_asset(name))
```

Reply:
Done and works, thx! Still, we have the minor type ignore when the checker doesn't know if is supposed for the function to raise instead of returning a `None`. So that would avoid:
```
error: Value of type "Optional[Asset]" is not indexable  [index]
```